src/Components/WebAssembly/WebAssembly/src/Services/LazyAssemblyLoader.cs

```diff
@@ -20,19 +19,18 @@ namespace Microsoft.AspNetCore.Components.WebAssembly.Services
     ///
     /// Supports finding pre-loaded assemblies in a server or pre-rendering context.
     /// </summary>
-    public class LazyAssemblyLoader
+    public sealed class LazyAssemblyLoader
     {
         internal const string GetDynamicAssemblies = "window.Blazor._internal.getLazyAssemblies";
         internal const string ReadDynamicAssemblies = "window.Blazor._internal.readLazyAssemblies";
 
-        private List<Assembly> _loadedAssemblyCache = new List<Assembly>();
+        private readonly IJSRuntime _jsRuntime;
+        private readonly Dictionary<string, Assembly> _loadedAssemblyCache;
 
-        private readonly IServiceProvider _provider;
-
-        public LazyAssemblyLoader(IServiceProvider provider)
+        public LazyAssemblyLoader(IJSRuntime jsRuntime)
         {
-            _provider = provider;
-            _loadedAssemblyCache = AppDomain.CurrentDomain.GetAssemblies().ToList();
+            _jsRuntime = jsRuntime;
+            _loadedAssemblyCache = AppDomain.CurrentDomain.GetAssemblies().ToDictionary(assembly => assembly.GetName().Name + ".dll", assembly => assembly);
```

Comment:
This is being super-defensive, `ToDictionary` will throw if if sees two items with the same key. In the unlikely event that GetAssemblies returns multiple assemblies with the same name, we would throw an unnecessary exception. We've been bitten by this in other places that deal with assemblies. If you manually construct the dictionary by setting the value, you can avoid the error.

```C#
_loadedAssemblyCache = new Dictionary<string, Assembly>(StringComparer.OrdinalIgnoreCase);
foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies())
{
  _loadedAssemblyCache[assembly.GetName().Name + ".dll"] = assembly;
}
```

Reply:
Yes, this actually happened in the tests. I've updated this now.